tests/chainer_tests/functions_tests/connection_tests/test_deconvolution_nd.py

```diff
@@ -52,91 +48,142 @@
-        self.gy = numpy.random.uniform(-1, 1, gy_shape).astype(self.x_dtype)
-
-        self.ggx = numpy.random.uniform(
-            -1, 1, self.x.shape).astype(self.x.dtype)
-        self.ggW = numpy.random.uniform(
-            -1, 1, self.W.shape).astype(self.W.dtype)
-        self.ggb = numpy.random.uniform(
-            -1, 1, self.b.shape).astype(self.b.dtype)
-
-        self.test_forward_options = {}
-        self.check_backward_options = {
-            'dtype': numpy.float64, 'atol': 3e-5, 'rtol': 3e-4}
+        self.x_shape = (self.N, self.in_channels) + self.dims
+        self.gy_shape = (self.N, self.out_channels) + outs
+
+        self.check_backward_options.update({'atol': 3e-5, 'rtol': 3e-4})
+        self.check_double_backward_options.update({'atol': 5e-3, 'rtol': 5e-2})
         if (self.x_dtype == numpy.float16 or self.W_dtype == numpy.float16
                 or self.b_dtype == numpy.float16):
-            self.test_forward_options = {'atol': 5e-4, 'rtol': 5e-3}
-            self.check_backward_options = {
-                'dtype': numpy.float64, 'atol': 2 ** -4, 'rtol': 2 ** -4}
-
-    def check_forward_consistency(self, use_cudnn='always'):
-        x_cpu = chainer.Variable(self.x)
-        W_cpu = chainer.Variable(self.W)
-        b_cpu = None if self.nobias else chainer.Variable(self.b)
+            self.check_forward_options.update({'atol': 5e-4, 'rtol': 5e-3})
+            self.check_backward_options.update({
+                'atol': 2 ** -4, 'rtol': 2 ** -4})
+            self.check_double_backward_options.update({
+                'atol': 2 ** -4, 'rtol': 2 ** -4})
+
+    def before_test(self, test_name):
+        self.backend_config.autotune = self.autotune
+
+    def generate_inputs(self):
+        W = numpy.random.normal(
+            0, self.W_scale, self.W_shape).astype(self.W_dtype)
+        x = numpy.random.uniform(-1, 1, self.x_shape).astype(self.x_dtype)
+        if self.nobias:
+            return x, W
+        else:
+            b = numpy.random.uniform(
+                -1, 1, self.out_channels).astype(self.x_dtype)
+            return x, W, b
+
+    def forward_expected(self, inputs):
+        if self.nobias:
+            x, W = inputs
+            b = None
+        else:
+            x, W, b = inputs
+        y_expected = F.deconvolution_nd(
+            x, W, b, stride=self.stride, pad=self.pad,
+            outsize=self.outsize, dilate=self.dilate,
+            groups=self.groups)
+        return y_expected.array,
+
+    def forward(self, inputs, device):
+        if self.nobias:
+            x, W = inputs
+            b = None
+        else:
+            x, W, b = inputs
+        y = F.deconvolution_nd(
+            x, W, b, stride=self.stride, pad=self.pad,
+            outsize=self.outsize, dilate=self.dilate,
+            groups=self.groups)
+        return y,
+
+    def check_forward_consistency(self, backend_config):
+        inputs = self.generate_inputs()
```

Comment:
Isnt this repeated from forward/forward_expected pair?
I think it should be possible to delete this as it is virtually doing the same than test forward

Reply:
Current `forward_expected` implementation depends on `F.deconvolution_nd` itself and thus it's only capable of checking consistency between backends, not absolute correctness of computations. Let's add a comment about it in `forward_expected` and let's remove `check_forward_consistency` tests.